internal/agent/device/hook/actions.go

```diff
@@ -1,248 +1,184 @@
+	for _, key := range []CommandLineVarKey{PathKey, FilesKey, CreatedKey, UpdatedKey, RemovedKey, BackupKey} {
+		actionCtx.commandLineVars[key] = ""
 	}
 }
 
-func (a *apiHookActionFactory) createExecutableActionHook(exec executer.Executer, log *log.PrefixLogger) (ActionHook, error) {
-	spec, err := a.AsHookAction0()
-	if err != nil {
-		return nil, err
-	}
-	actionTimeout, err := parseTimeout(spec.Executable.Timeout)
-	if err != nil {
-		return nil, err
+func computeFileDiff(actionCtx *actionContext, current *api.RenderedDeviceSpec, desired *api.RenderedDeviceSpec) {
+	currentFileList := []ignv3types.File{}
+	desiredFileList := []ignv3types.File{}
+
+	if current != nil && current.Config != nil {
+		currentIgnition, err := config.ParseAndConvertConfigFromStr(*current.Config)
+		if err == nil {
+			currentFileList = append(currentFileList, currentIgnition.Storage.Files...)
+		}
 	}
-	envVars := util.FromPtr(spec.Executable.EnvVars)
-	if err = validateEnvVars(envVars); err != nil {
-		return nil, err
+	if desired != nil && desired.Config != nil {
+		desiredIgnition, err := config.ParseAndConvertConfigFromStr(*desired.Config)
+		if err == nil {
+			desiredFileList = append(desiredFileList, desiredIgnition.Storage.Files...)
+		}
 	}
-	return newExecutableActionHook(spec.Executable.Run,
-		util.FromPtr(spec.Executable.EnvVars),
-		exec,
-		actionTimeout,
-		spec.Executable.WorkDir,
-		log), nil
-}
 
-func (a *apiHookActionFactory) createSystemdActionHook(exec executer.Executer, log *log.PrefixLogger) (ActionHook, error) {
-	spec, err := a.AsHookAction1()
-	if err != nil {
-		return nil, err
+	currentFileMap := make(map[string]ignv3types.File)
+	for _, f := range currentFileList {
+		currentFileMap[f.Path] = f
 	}
-	actionTimeout, err := parseTimeout(spec.Systemd.Timeout)
-	if err != nil {
-		return nil, err
+	for _, f := range desiredFileList {
+		if content, ok := currentFileMap[f.Path]; !ok {
+			actionCtx.createdFiles[f.Path] = ignv3types.File{}
+		} else if !reflect.DeepEqual(f, content) {
+			actionCtx.updatedFiles[f.Path] = ignv3types.File{}
+		}
 	}
-	return newSystemdActionHook(spec.Systemd.Unit.Name,
-		spec.Systemd.Unit.Operations,
-		exec,
-		actionTimeout,
-		log), nil
-}
 
-func (a *apiHookActionFactory) Create(exec executer.Executer, log *log.PrefixLogger) (ActionHook, error) {
-	hookActionType, err := a.Type()
-	if err != nil {
-		return nil, err
+	desiredFileMap := make(map[string]ignv3types.File)
+	for _, f := range desiredFileList {
+		desiredFileMap[f.Path] = f
 	}
-	var actionHook ActionHook
-	switch hookActionType {
-	case v1alpha1.ExecutableActionType:
-		actionHook, err = a.createExecutableActionHook(exec, log)
-	case v1alpha1.SystemdActionType:
-		actionHook, err = a.createSystemdActionHook(exec, log)
-	default:
-		return nil, fmt.Errorf("%w: %s", errors.ErrActionTypeNotFound, hookActionType)
+	for _, f := range currentFileList {
+		if content, ok := desiredFileMap[f.Path]; !ok {
+			actionCtx.removedFiles[f.Path] = content
```

Comment:
I get why we are doing this but do we want to maintain the entire file in memory?